kv,sql: transaction commit deadlines due to leases should be late-bound

fqazi opened the pull request:
Previously, SQL statements, during their execution, lease
tables from the lease.Manager indirectly through the
connExecutor's descs.Collection. Whenever a lease is used,
the deadline of that version of that lease is applied to
the transaction as a commit deadline. This had the unfortunate
side effect that renewals would not be properly picked up, so
if a transaction took longer then the 4 to 5 minute validity
period, then we can get into problematic scenarios.

To address, this is patch, adds lease tracking through the
the table collections, so that we are aware when extension
occurs, and uses that to figure out the latest time a given
transaction can commit.

Fixes #51042

Release note (bug fix): Fixed a bug that prevent transactions which lasted
longer than 5 minutes and then performed writes from committing.

fqazi committed:
kv,sql: transaction commit deadlines due to leases should be late-bound

Previously, SQL statements, during their execution, lease
tables from the lease.Manager indirectly through the
connExecutor's descs.Collection. Whenever a lease is used,
the deadline of that version of that lease is applied to
the transaction as a commit deadline. This had the unfortunate
side effect that renewals would not be properly picked up, so
if a transaction took longer then the 4 to 5 minute validity
period, then we can get into problematic scenarios.

To address, this is patch, adds lease tracking through the
the table collections, so that we are aware when extension
occurs, and uses that to figure out the latest time a given
transaction can commit.

Fixes #51042

Release note (bug fix): Fixed a bug that prevent transactions which lasted
longer than 5 minutes and then performed writes from committing.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63725)
<!-- Reviewable:end -->

knz commented:
Tobias suggested me to use this PR as an experiment to test our new notification system.
Thus: cc @cockroachdb/kv

ajwerner reviewed (commented):
Sorry for the delayed review. Two big things:

I think we need to have clearer principles regarding where we update the deadline in the conn_executor. 

It'd also be nice to have a test that exercises the renewal by shortening the durations dramatically and showing that this patch has the intended impact on the ability of a transaction to observe a refreshed lease and then get pushed beyond an original expiration.

> Reviewed 18 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-:-MYroofS20xiwFOm3b1N:b94t5tf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[pkg/kv/txn.go, line 688 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrlh4J4EdhloFYwOCP:-MYrlh4J4EdhloFYwOCQ:b41g8s8) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/kv/txn.go#L688)):*
> ```Go
> 
> // UpdateDeadline sets the transactions deadline to the passed deadline.
> // It may move the deadline further into the future.
> ```

it also may move it into the past. How about `It may move the deadline to any timestamp above the current read timestamp.` Maybe also add a note that it's probably weird to set it to something below the current provisional commit timestamp but that that isn't enforced here. 
___
*[pkg/kv/txn.go, line 702 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrlV2lCaicVyu6ARyo:-MYrlV2mCqpf4Hh5jwVr:b-8xld6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/kv/txn.go#L702)):*
> ```Go
> 	if deadline.Less(readTimestamp) {
> 		log.Fatalf(ctx, "deadline below read timestamp is nonsensical; "+
> 			"txn has would have no change to commit. Deadline: %s. Read timestamp: %s Previous Deadline: %s.",
> ```

nit: you didn't write it but can we spruce up this message. `s/change/chance/` and then lose the upper-case word beginnings.
___
*[pkg/settings/setting.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrmQTR-O7ojkqx03gz:-MYrmQTR-O7ojkqx03h-:bm693oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/settings/setting.go#L23)):*
> ```Go
> // MaxSettings is the maximum number of settings that the system supports.
> // Exported for tests.
> const MaxSettings = 259
> ```

meh, go to 512. 
___
*[pkg/sql/conn_executor.go, line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrmZ5033iR6mKoOHBC:-MYrmZ5033iR6mKoOHBD:b-q2p6ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/sql/conn_executor.go#L1500)):*
> ```Go
> 	case ExecStmt:
> 		// Update the deadline on the transaction based on the collections
> 		if deadline, haveDeadline := ex.extraTxnState.descCollection.Deadline(); haveDeadline && ex.state.mu.txn != nil {
> ```

I think this should happen in the state transition itself rather than here. The idea is that the state of the transaction changes, generally, underneath the state machine. This code is really hard to get right and not as principled as it ought to be. It may take some iteration to get this part clean. I think I was a lot less aware of the design of this machine back then. 
___
*[pkg/sql/conn_executor_exec.go, line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYroeE51A76rA8Jq_M0:-MYroeE51A76rA8Jq_M1:b-lpx8dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/sql/conn_executor_exec.go#L254)):*
> ```Go
> 	var stmt Statement
> 	queryID := ex.generateID()
> 	// Update the deadline on the transaction based on the collections
> ```

nit: period. 
___
*[pkg/sql/conn_executor_prepare.go, line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrnBCo0cxlXBh-aBfE:-MYrnBCo0cxlXBh-aBfF:b-jbwi5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/sql/conn_executor_prepare.go#L229)):*
> ```Go
> 			return nil, err
> 		}
> 		// Prepare is an implicit transaction and we know it with any
> ```

I could see pulling this out into its own PR or commit. This is a very subtle correctness issue. 
___
*[pkg/sql/catalog/lease/lease.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrntE0ENMJJ5p114he:-MYrntE0ENMJJ5p114hf:b-tb2tnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/sql/catalog/lease/lease.go#L66)):*
> ```Go
> func between0and1inclusive(f float64) error {
> 	if f < 0 || f > 1 {
> 		return errors.Errorf("must be between 0 and 1")
> ```

nit: you've got the `f` there, may as well tell us the provided value. 


<!-- Sent from Reviewable.io -->

fqazi committed:
sql: implicit transactions did not properly release leases

Previously, when an implicit transaction was used during
the prepare phase a new transaction is constructed, which
will acquire leases as needed. The descCollections will be
shared between the execution and prepare phases, which could
lead to these leases being held during the execute phase even
though the prepare transaction is complete. To address this
the current patch will release those leases right after the
prepare phase.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/settings/setting.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrmQTR-O7ojkqx03gz:-MYuqYCu9NyA4rsA4QdW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/settings/setting.go#L23)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

meh, go to 512. 
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrmZ5033iR6mKoOHBC:-MYuqZAeDnppfICkxynA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/sql/conn_executor.go#L1500)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this should happen in the state transition itself rather than here. The idea is that the state of the transaction changes, generally, underneath the state machine. This code is really hard to get right and not as principled as it ought to be. It may take some iteration to get this part clean. I think I was a lot less aware of the design of this machine back then. 
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_prepare.go, line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYrnBCo0cxlXBh-aBfE:-MYuq_vO8mx5X4vkvNfk:b12sag3) ([raw file](https://github.com/cockroachdb/cockroach/blob/439efecd3bfd956ac2e8990f46e1cc1eed24a2b5/pkg/sql/conn_executor_prepare.go#L229)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I could see pulling this out into its own PR or commit. This is a very subtle correctness issue. 
</blockquote></details>

I'll make this its own commit, since it shouldn't get lost.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
CI looks unhappy, any clue why? A rebase is in order too. 

> Reviewed 2 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-:-MYvrQbv24oZ4f3MiLT_:byazllu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/catalog/lease/lease_test.go, line 2879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYvqckX9I3UNCpLztR-:-MYvqckX9I3UNCpLztR0:b-avrbcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/53891bd76c14032be78063f3ed9e2dbdce738a44/pkg/sql/catalog/lease/lease_test.go#L2879)):*
> ```Go
> SELECT * FROM t1;
> select pg_sleep(5);
> INSERT INTO t1 VALUES (1);
> ```

This works for an interesting reason. You don't get auto-pushed in cockroach to the present when writing. There is a "closed timestamp" that trails the present that will push writes. The default closed timestamp trails the present by roughly 3 seconds. You could shorten that. The problem with it is that it is a best-effort thing which means that this could get flakey. If you want to ensure that the transaction gets pushed, you'll need to use a separate transaction to read the key. 

I could be convinced that relying on the closed timestamp is okay. One option to ensure that the closed timestamp has occurred is by creating a rangefeed over the span. That will send you a checkpoint telling you when a closed timestamp has passed. 


<!-- Sent from Reviewable.io -->

fqazi commented:
> CI looks unhappy, any clue why? A rebase is in order too.
> 
> > Reviewed 2 of 7 files at r2.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-:-MYvrQbv24oZ4f3MiLT_:byazllu)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)
> 
> _[pkg/sql/catalog/lease/lease_test.go, line 2879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MYvqckX9I3UNCpLztR-:-MYvqckX9I3UNCpLztR0:b-avrbcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/53891bd76c14032be78063f3ed9e2dbdce738a44/pkg/sql/catalog/lease/lease_test.go#L2879)):_
> 
> > ```go
> > SELECT * FROM t1;
> > select pg_sleep(5);
> > INSERT INTO t1 VALUES (1);
> > ```
> 
> This works for an interesting reason. You don't get auto-pushed in cockroach to the present when writing. There is a "closed timestamp" that trails the present that will push writes. The default closed timestamp trails the present by roughly 3 seconds. You could shorten that. The problem with it is that it is a best-effort thing which means that this could get flakey. If you want to ensure that the transaction gets pushed, you'll need to use a separate transaction to read the key.
> 
> I could be convinced that relying on the closed timestamp is okay. One option to ensure that the closed timestamp has occurred is by creating a rangefeed over the span. That will send you a checkpoint telling you when a closed timestamp has passed.

Let me dig into the failures didn't notice these locally. Not clear from the logs, need to dig deeper, since the connection mysteriously dies. Have a theory if it is linked to the changes.

fqazi commented:
Linked to the transaction start not having deadline assertion, let me dig into what the root cause.

andreimatei reviewed (commented):
@fqazi , since you're doing work in the area, can I kindly ask you to look into #24020 and see if  the weirdness with rolling back a txn that exceeded its deadline is still current? The issue has been closed, but for no good reason.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-:-MYzDIm0DyROuwxC0ewy:b-b0my40)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 4 of 21 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/conn_executor.go, line 2337 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLuQXzBzcroUbO4Z5z:-MZLuQXzBzcroUbO4Z6-:bo0wkg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/conn_executor.go#L2337)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
> 		if txnIsOpen {
> 			if deadline, haveDeadline := ex.extraTxnState.descCollection.Deadline(); haveDeadline && ex.state.mu.txn != nil {
> 				err := ex.state.mu.txn.UpdateDeadline(ex.Ctx(), deadline)
> 				if err != nil {
> 					return advanceInfo{}, err
> 				}
> 			}
> 		}
> ```
</details>

nit: pull this into a helper `maybeUpdateDeadline` or something like that. 
___
*[pkg/sql/catalog/lease/lease.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLyTpO5e1QmB6W92GN:-MZLyTpO5e1QmB6W92GO:bf53u8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease.go#L79)):*
> ```Go
> 
> // LeaseRenewalDuration controls the default time before a lease expires when
> //// acquisition to renew the lease begins.
> ```

nit: extra `//`
___
*[pkg/sql/catalog/lease/lease_test.go, line 2854 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLukgc3wpm6xE0MJjV:-MZLukgc3wpm6xE0MJjW:b-f6pxua) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2854)):*
> ```Go
> // Validates that the transaction deadline can be extended
> // past the original lease duration. Previously, we had a
> // a limitation if the transaction took longer then this
> ```

incomplete sentence.
___
*[pkg/sql/catalog/lease/lease_test.go, line 2874 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLvDb5-j33FCZX44PU:-MZLvDb68Knr20tRmTye:b6a3ngr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2874)):*
> ```Go
> 			// it to resume when the channel gets unblocked.
> 			if req.Txn != nil && req.Txn.ID.String() == txnID {
> 				filterMu.Unlock()
> ```

it's a little subtle that this doesn't need locking on `blockedOnce` because it assumes there will be at most one inflight `EndTxn` for the transaction. Comment?
___
*[pkg/sql/catalog/lease/lease_test.go, line 2904 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLvU1S-Rrjmccvz01O:-MZLvU1S-Rrjmccvz01P:bl9jgf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2904)):*
> ```Go
> BEGIN;
> SELECT * FROM t1;
> select pg_sleep(5);
> ```

the issue I have with this test is that it's slow. I think I'd love to see this work in a very similar way to the below test except without the schema change. 
___
*[pkg/sql/catalog/lease/lease_test.go, line 2921 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLvoWU7InZmbxpnnGm:-MZLvoWU7InZmbxpnnGn:blh5fpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2921)):*
> ```Go
> 		resumeChan := make(chan struct{})
> 
> 		//
> ```

missing comment body
___
*[pkg/sql/catalog/lease/lease_test.go, line 2935 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLvyA510sQBXPF1I9v:-MZLvyA510sQBXPF1I9w:b-1p84mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2935)):*
> ```Go
> 			}
> 			// Fetch the transaction ID, so that we can delay the commit
> 			txnIDResult, err := conn.QueryContext(ctx, `SELECT id FROM crdb_internal.node_transactions WHERE session_id IN (SELECT * FROM [SHOW session_id]);`)
> ```

This seems like a good use of `QueryRowContext`
___
*[pkg/sql/catalog/lease/lease_test.go, line 2948 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLw7ZX-0zhysHFnxMX:-MZLw7ZX-0zhysHFnxMY:bt12uv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2948)):*
> ```Go
> 				return
> 			}
> 			require.NoError(t, err)
> ```

I don't think you want to call `require` on a goroutine. Tests can deadlock if `t.Fatal` gets called on a separate goroutine.
___
*[pkg/sql/catalog/lease/lease_test.go, line 2957 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLx7CV4Eq6TdFNSDXS:-MZLx7CV4Eq6TdFNSDXT:bd7qfch) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2957)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 			// Execute an insert on the same connection and attempt
> 			// to commit, this operation will fail.
> 			_, err = conn.ExecContext(ctx, `
> INSERT INTO t1 VALUES (1);
> ```
</details>

I think it's worth breaking this up. 100% of the time this should fail on the `INSERT`. I say that because by the time we do that insert, our deadline should be set to some timestamp below we need to be pushed in order to not rewrite history. My guess is that in the case where we hit the request filter above, we're aborting the transaction and not committing it. 

All that is to say, I'm not sure we need the request filter to exercise the case where the renewal does not happen. 

What I would use the RequestFilter for is to ensure that indeed the deadline does move in the case where there is no schema change. I might do that by just pulling the deadline out of each and every require corresponding to the txn and putting it in some slice or something and then making sure it goes up. 
___
*[pkg/sql/catalog/lease/lease_test.go, line 2974 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLwPjr3HF-sTihbdeJ:-MZLwPjr3HF-sTihbdeK:b-ur1jlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2974)):*
> ```Go
> 		require.NoError(t, err)
> 		// Give the lease time to expire
> 		time.Sleep(time.Second * 5)
> ```

I thought the leases had a 0 duration; does it need any time to expire? 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/catalog/lease/lease_test.go, line 2957 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLx7CV4Eq6TdFNSDXS:-MZNohKoDpoPHEWPTFVm:b-c98gaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2957)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 			// Execute an insert on the same connection and attempt
> 			// to commit, this operation will fail.
> 			_, err = conn.ExecContext(ctx, `
> INSERT INTO t1 VALUES (1);
> ```
I think it's worth breaking this up. 100% of the time this should fail on the `INSERT`. I say that because by the time we do that insert, our deadline should be set to some timestamp below we need to be pushed in order to not rewrite history. My guess is that in the case where we hit the request filter above, we're aborting the transaction and not committing it. 

All that is to say, I'm not sure we need the request filter to exercise the case where the renewal does not happen. 

What I would use the RequestFilter for is to ensure that indeed the deadline does move in the case where there is no schema change. I might do that by just pulling the deadline out of each and every require corresponding to the txn and putting it in some slice or something and then making sure it goes up. 
</blockquote></details>

Ignore me about the insert failing. I was thinking the transaction was carrying its deadline everywhere with it. One interesting thing is that by the time the insert finishes, the new deadline will be below the write timestamp. It's not below the read timestamp only because we don't eagerly refresh. That's quite subtle. I'm going to file an issue. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/catalog/lease/lease_test.go, line 2854 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLukgc3wpm6xE0MJjV:-MZNdbuZ2_e63uLNlhTY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2854)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

incomplete sentence.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/lease_test.go, line 2874 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLvDb5-j33FCZX44PU:-MZNdfDX2Pt6Fv6v383D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2874)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it's a little subtle that this doesn't need locking on `blockedOnce` because it assumes there will be at most one inflight `EndTxn` for the transaction. Comment?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/lease_test.go, line 2921 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLvoWU7InZmbxpnnGm:-MZNe7mOEsOB6KD8_Eyt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2921)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

missing comment body
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/lease_test.go, line 2948 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLw7ZX-0zhysHFnxMX:-MZNejzcDoxrnq94gvfE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2948)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think you want to call `require` on a goroutine. Tests can deadlock if `t.Fatal` gets called on a separate goroutine.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/lease_test.go, line 2957 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLx7CV4Eq6TdFNSDXS:-MZNqNhxDd0wgHDHbB8A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2957)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Ignore me about the insert failing. I was thinking the transaction was carrying its deadline everywhere with it. One interesting thing is that by the time the insert finishes, the new deadline will be below the write timestamp. It's not below the read timestamp only because we don't eagerly refresh. That's quite subtle. I'm going to file an issue. 
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/lease_test.go, line 2974 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZLwPjr3HF-sTihbdeJ:-MZNqTfSEAyZWHh6ZZq8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c09098ba9b2099e74111425018dfaad6877a2b9/pkg/sql/catalog/lease/lease_test.go#L2974)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I thought the leases had a 0 duration; does it need any time to expire? 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, just minor things left. I'd like to understand the `nil` transaction cases better. 

> Reviewed 10 of 21 files at r4, 2 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-:-MZNwcUR8l7fqk-XL5cJ:bd9p08j)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/kv/txn.go, line 690 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNuzV39f1F7scSAP7J:-MZNuzV39f1F7scSAP7K:b-9ty0lt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/kv/txn.go#L690)):*
> ```Go
> // It may move the deadline to any timestamp above the current read timestamp.
> //
> // The deadline cannot be lower than txn.ReadTimestamp.
> ```

This ends up being rather subtle so a bit more commentary would be appreciated. Note that if the deadline is below the current provisional commit timestamp (write timestamp), nothing happens but that that transaction is doomed. I might also add some commentary that we are relying on the read timestamp not changing during execution in order to uphold this invariant. That is valid today but it's fragile.
___
*[pkg/sql/conn_executor.go, line 2337 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNvNkSE_QHBmHZDd9I:-MZNvNkTD0gRbfoBPcnn:b-r97805) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/sql/conn_executor.go#L2337)):*
> ```Go
> 		// Update the deadline on the transaction based on the collections,
> 		// if the transaction is currently open.
> 		if txnIsOpen && ex.state.mu.txn != nil {
> ```

When are we in the open state but `txn` is nil?
___
*[pkg/sql/conn_executor.go, line 2337 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNvapY674BUTBRwXFb:-MZNvapY674BUTBRwXFc:b17eqzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/sql/conn_executor.go#L2337)):*
> ```Go
> 		// Update the deadline on the transaction based on the collections,
> 		// if the transaction is currently open.
> 		if txnIsOpen && ex.state.mu.txn != nil {
> ```

nit: Would it be crazy to add yet one more layer of indirection here by creating `ex.maybeUpdateDeadline` which checks that the transaction is open and then updates the deadline?
___
*[pkg/sql/conn_executor_prepare.go, line 230 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNvug57Iv5guvzUScM:-MZNvug69mTzUMToBVak:b-nrrmqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/sql/conn_executor_prepare.go#L230)):*
> ```
> 		// Prepare is an implicit transaction and we know it with any
> 		// other descriptors. Once the implicit transaction is done
> ```
Something got garbled in this sentence.


<!-- Sent from Reviewable.io -->

fqazi commented:
Thanks for the quick review @ajwerner! Let me check when those cases occur

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/kv/txn.go, line 690 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNuzV39f1F7scSAP7J:-MZO7Rwh8wiYwfpTxxvc:b-1qcfmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/kv/txn.go#L690)):*
> This ends up being rather subtle so a bit more commentary would be appreciated. Note that if the deadline is below the current provisional commit timestamp (write timestamp), nothing happens but that that transaction is doomed. I might also add some commentary that we are relying on the read timestamp not changing during execution in order to uphold this invariant. That is valid today but it's fragile.

Done changed to:
// UpdateDeadline sets the transactions deadline to the passed deadline.
// It may move the deadline to any timestamp above the current read timestamp.
// If the deadline is below the current provisional commit timestamp (write timestamp),
// then the transaction will fail with a deadline error during the commit.
// The deadline cannot be lower than txn.ReadTimestamp and we make the assumption
// the read timestamp will not change during execution, which is valid today.
___
*[pkg/sql/conn_executor_prepare.go, line 230 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNvug57Iv5guvzUScM:-MZO6QOZDz0bPnYQJle1:b6r33xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/sql/conn_executor_prepare.go#L230)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		// Prepare is an implicit transaction and we know it with any
> 		// other descriptors. Once the implicit transaction is done
> ```
Something got garbled in this sentence.
</blockquote></details>

Done:		
                 // Prepare with an implicit transaction will end up creating
		// a new transaction. Once this transaction is complete,
		// we can safely release the leases, otherwise we will
		// incorrectly hold leases for later operations.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[pkg/sql/conn_executor.go, line 2337 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNvNkSE_QHBmHZDd9I:-MZOmS0A8XhVejVC108v:b-be31pq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/sql/conn_executor.go#L2337)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

When are we in the open state but `txn` is nil?
</blockquote></details>

Removed and didn't run into any broken scenario
___
*[pkg/sql/conn_executor.go, line 2337 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-MZNvapY674BUTBRwXFb:-MZOmVK53Mv9UkLMjUQY:b-8o7rth) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcda1f3cba98f5001aa38c2a3ece76744568cba/pkg/sql/conn_executor.go#L2337)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: Would it be crazy to add yet one more layer of indirection here by creating `ex.maybeUpdateDeadline` which checks that the transaction is open and then updates the deadline?
</blockquote></details>

Discarding this one, since only one place checks the open state now.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63725#-:-MZPTFqJ16l1HAcHqVqA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

fqazi commented:
bors r=ajwerner
